### PR TITLE
pollKey implementation for w32 backend of highgui

### PR DESCRIPTION
**Pull request by crackwitz**
first code proposal for #19410

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Win32,Custom
buildworker:Custom=linux-1,linux-4,linux-6
build_image:Custom=qt:16.04
```

### CODE REVIEW

File: modules/highgui/src/window_winrt.cpp [hunk 1]
```diff
@@ -205,6 +205,18 @@ CV_IMPL int cvGetTrackbarPos(const char* trackbar_name, const char* window_name)
...
+CV_IMPL int cvPollKey()
+{
+    CV_WINRT_NO_GUI_ERROR("cvPollKey");
+
+    for (;;)
+    {
+        CvWindow* window;
+
+        // TODO: implement appropriate logic here
+    }
```
Comment (asmorkalov) [inline modules/highgui/src/window_winrt.cpp:217]: It's obviously dead code.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/highgui/src/window_winrt.cpp (hunk 1 above)
Comment (crackwitz) [reply modules/highgui/src/window_winrt.cpp]: loop removed, comment remains as in adjacent functions.

File: modules/highgui/include/opencv2/highgui/highgui_c.h [hunk 2]
```diff
@@ -210,6 +210,9 @@ CVAPI(void) cvSetMouseCallback( const char* window_name, CvMouseCallback on_mous
 /* wait for key event infinitely (delay<=0) or for "delay" milliseconds */
 CVAPI(int) cvWaitKey(int delay CV_DEFAULT(0));
 
+/* poll for key event */
+CVAPI(int) cvPollKey(void);
```
[the remark this answers is not in the document]
Comment (crackwitz) [reply modules/highgui/include/opencv2/highgui/highgui_c.h]: reverted.

**Comment by asmorkalov** [comment]
@crackwitz Please pay attention on CI. The PR breaks build.

**Comment by crackwitz** [comment]
checking.

**Comment by crackwitz** [comment]
how should I handle this? all the existing code for every function is structured such that there is an extern C implementation of the respective function per toolkit, and then one namespaced wrapper function calling that.

I could remove the CV_IMPL (extern C) from all implementations of cvPollKey so it iself is not exposed. I think that will then require a forward declaration in `window.cpp` before/inside of `cv::pollKey` since the C function isn't declared in a common header anymore.

**Comment by crackwitz** [comment]
ok, looks like I did it right this time. I hope the structure is as desired.

**Review by asmorkalov — APPROVED** [review]
Well done! :+1:

**Comment by asmorkalov** [comment]
@crackwitz Please squash commits to have clean changes history in OpenCV git.

File: modules/highgui/include/opencv2/highgui.hpp [hunk 3]
```diff
@@ -364,6 +364,24 @@ If there are several HighGUI windows, any of them can be active.
...
+/** @brief Polls for a pressed key.
+
+The function pollKey polls for a key event without waiting. It returns the code of the pressed key
+or -1 if no key was pressed since the last invocation.
+
+@note
+
+This function serves a similar function as waitKey. Either function needs to be
+called periodically for normal event processing unless HighGUI is used within an environment that
+takes care of event processing.
```
Comment (alalek) [inline modules/highgui/include/opencv2/highgui.hpp:376]: This note is [malformed](http://pullrequest.opencv.org/buildbot/export/pr/19411/docs/d7/dfc/group__highgui.html#ga6d20fbd3100ec3badc1eaa653aff99d7).

Need to remove new lines.

File: modules/highgui/src/window_w32.cpp [hunk 4]
```diff
@@ -1981,6 +1981,97 @@ static void showSaveDialog(CvWindow* window)
     }
 }
 
+/*
+ * message received. check if it belongs to our windows (frame, hwnd).
+ * returns true (and value in result) if there was a keycode.
+ * otherwise return false (keep getting messages).
+ */
+static bool handleMessage(MSG& message, int& result)
```
Comment (alalek) [inline modules/highgui/src/window_w32.cpp:1989]: > result

`keyCode`

**Review by alalek — COMMENTED** [review]
Please add/update documentation's cross-links between waitKey(), imshow() calls. So Users can find this functionality.

BTW, This function's purpose is updating UI too (like `waitKey()`)

(function's name is still confusing to me, perhaps @vpisarev can suggest something or approve this)

File: modules/highgui/include/opencv2/highgui.hpp (hunk 3 above)
Comment (crackwitz) [reply modules/highgui/include/opencv2/highgui.hpp:376]: I don't quite understand. remove Line 373 which is a blank line between `@note` and its body? or reformat paragraph to use 72-80 characters of width consistently?

File: modules/highgui/src/window_w32.cpp (hunk 4 above)
Comment (crackwitz) [reply modules/highgui/src/window_w32.cpp:1989]: I'll rename it and the corresponding instances in waitKey and pollKey

**Comment by crackwitz** [comment]
here are the requested changes. I hope I understood the intention.

I'll squash when you're satisfied with the state of the pull request. squashing implies force-push; I hope that's correct.

**Comment by crackwitz** [comment]
I've squashed the commits so far and rebased to current master HEAD.

**Review by alalek — APPROVED** [review]
Thank you!

**Pull request merged**